### PR TITLE
Scheduling algorithm

### PR DESCRIPTION
**Pull request by epeterson42**
Add scheduling algorithm code to the main powerplay project

### CODE REVIEW

File: backend/internal/server/apis/schedule/games/games.go [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+	log.Info("Pre-optimization analysis")
+	seasonStats, teamStats := analysis.RunTimeAnalysis(games)
+
+	// Need to make sure games are balanced in
+	// - Early / late
+	// - Days between games
+	balanceCount := getBalanceCount(&teamStats)
+	lastBalanceCount := -1
+
+	for count := 0; balanceCount != lastBalanceCount && count < 25; count++ {
```
Comment (marcus-quincy) [inline backend/internal/server/apis/schedule/games/games.go:108]: One thing I have seen happen when playing with the scheduler is sometimes when running the optimization multiple times a subsequent run will be less balanced than the previous one. So an alternative to this logic would be to run it multiple times and pick the one that's most balanced

File: backend/internal/server/apis/schedule/pkg/analysis/analysis.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+		}
+
+		stats.AverageDaysBetweenGames /= float32(len(stats.DaysBetweenGames))
+
+		teamStats[team] = stats
+	}
+}
+
+func printStats(seasonStats models.SeasonStats, teamStats map[string]models.TeamStats) {
+	log.Info("Early games: %v/%v (%v%%)\n", seasonStats.EarlyGames, seasonStats.TotalGames, (float32(seasonStats.EarlyGames)/float32(seasonStats.TotalGames))*100)
```
Comment (marcus-quincy) [inline backend/internal/server/apis/schedule/pkg/analysis/analysis.go:104]: For the analysis do we want it printed out to the logs? It seems like we would want it output separately from the rest of the logs, where it is potentially interwoven with other log statements like user requests

File: backend/internal/server/apis/schedule/pkg/parser/games.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+	err = gocsv.UnmarshalFile(scheduleFile, &games)
+	if err != nil {
+		log.Error("Failed to unmarshal schedule file: %v\n", err)
+	}
+
+	log.Info("Read %v games\n", len(games))
+
+	for i := range games {
+		games[i].Start, err = time.Parse("01/02/2006 15:04", fmt.Sprintf("%v %v", games[i].StartDate, games[i].StartTime))
```
Comment (marcus-quincy) [inline backend/internal/server/apis/schedule/pkg/parser/games.go:36]: I am confused why we are parsing this string into a time?

**Review by marcus-quincy — COMMENTED** [review]
Overall things look OK i just have some questions.

File: backend/internal/server/apis/schedule/games/games.go (hunk 1 above)
Comment (epeterson42) [reply backend/internal/server/apis/schedule/games/games.go:108]: That's a good idea. Think we should include it in this PR or do it as another feature?

File: backend/internal/server/apis/schedule/pkg/analysis/analysis.go (hunk 2 above)
Comment (epeterson42) [reply backend/internal/server/apis/schedule/pkg/analysis/analysis.go:104]: Do you have any thoughts on how to separate it? Maybe we return the analysis logs to the user or save them in the db attached to the schedule model?

File: backend/internal/server/apis/schedule/pkg/parser/games.go (hunk 3 above)
Comment (epeterson42) [reply backend/internal/server/apis/schedule/pkg/parser/games.go:36]: The first argument of time.Parse is a format string, is that what you are asking about?

Comment (zanehirning) [reply backend/internal/server/apis/schedule/pkg/parser/games.go:36]: > I am confused why we are parsing this string into a time?

Our current game model requires Start to be a time.Time, that is why we are parsing it. I think the idea is that we won't have the parser package when we are done? I'm not sure though

File: backend/internal/server/apis/schedule/games/games.go [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+		lastBalanceCount = balanceCount
+		balanceCount := getBalanceCount(&teamStats)
+
+		log.Info("Balanced count: %v\n", balanceCount)
+	}
+}
+
+func generateGames(leagues []models.League, numberOfGamesPerTeam int) (models.Season, error) {
+	if len(leagues) == 0 {
+		return models.Season{}, errors.New("no leagues to games games for")
```
Comment (zanehirning) [inline backend/internal/server/apis/schedule/games/games.go:123]: Kind of nitpicky, but this error isn't coherent

File: backend/internal/server/apis/schedule/pkg/analysis/analysis.go (hunk 2 above)
Comment (zanehirning) [reply backend/internal/server/apis/schedule/pkg/analysis/analysis.go:104]: For the most part, I wonder if we can just get rid of printStats. I think it was just a way for him to see how well his scheduling algorithm did, but I am not 100%

**Review by zanehirning — APPROVED** [review]
I think this is a good starting point. We have some things that won't be necessary to have, like parser and writing to the csv, etc.

File: backend/internal/server/apis/schedule/pkg/parser/games.go (hunk 3 above)
Comment (marcus-quincy) [reply backend/internal/server/apis/schedule/pkg/parser/games.go:36]: no  I am just confused about the significance of 01/02/2006 15:04

File: backend/internal/server/apis/schedule/games/games.go (hunk 1 above)
Comment (marcus-quincy) [reply backend/internal/server/apis/schedule/games/games.go:108]: separate feature is OK with me.

File: backend/internal/server/apis/schedule/pkg/parser/games.go (hunk 3 above)
Comment (marcus-quincy) [reply backend/internal/server/apis/schedule/pkg/parser/games.go:36]: Oh I guess go just always uses this date to specify the format. https://gobyexample.com/time-formatting-parsing

never seen a format string specified that way but i kinda like it

File: backend/internal/server/apis/schedule/games/games.go (hunk 1 above)
Comment (epeterson42) [reply backend/internal/server/apis/schedule/games/games.go:108]: https://github.com/jak103/powerplay/issues/118

File: backend/internal/server/apis/schedule/pkg/analysis/analysis.go (hunk 2 above)
Comment (epeterson42) [reply backend/internal/server/apis/schedule/pkg/analysis/analysis.go:104]: How does everyone feel about changing the log level to debug for now then if we want them later we can do something else with it?

File: backend/internal/server/apis/schedule/games/games.go (hunk 4 above)
Comment (epeterson42) [reply backend/internal/server/apis/schedule/games/games.go]: Fixed:)

**Comment by epeterson42** [comment]
We will reopen this when we get some things straightened out.

**Pull request closed without merging**